### DISCUSSION TITLE
How to use dictionaries as state variables

### DISCUSSION THREAD

**Discussion by EZoni** [Kitware/trame#599] [Q&A] [answered]
Hi,

I have two questions about how to use a dictionary as state variable in a Trame application.

Here's my code:
```py
import numpy as np

from trame.app import get_server
from trame.ui.vuetify2 import SinglePageLayout
from trame.widgets import vuetify2 as v2

# toy model
def model(x_array):
    return np.sum(x_array)

server = get_server(client_type="vue2")
state, ctrl = server.state, server.controller

# declare inputs
state.params = dict()
state.dirty("params")

# initialize inputs
state.params["a"] = 1
state.params["b"] = 2

# initialize output
params_array = np.array(list(state.params.values()))
state.output = model(params_array)

# update output
@state.change("params")
def params_react(params, **kwargs):
    params_array = np.array(list(params.values()))
    print(f"\nparams_array = {params_array}")
    print(f"old output = {state.output}")
    state.output = model(params_array)
    print(f"new output = {state.output}")

with SinglePageLayout(server) as layout:
    with layout.content:
        with v2.VContainer():
            with v2.VRow():
                with v2.VCol():
                    with v2.VCard(style="width: 500px"):
                        with v2.VCardTitle("Parameters"):
                            with v2.VCardText():
                                for key in state.params.keys():
                                    with v2.VSlider(
                                        v_model=(f"params['{key}']",),
                                        label=f"{key}",
                                        min=0,
                                        max=10,
                                        step=1,
                                        hide_details=True,
                                        classes="align-center",
                                    ):
                                        with v2.Template(
                                            v_slot_append=True,
                                            __properties=[("v_slot_append", "v-slot:append")],
                                        ):
                                            v2.VTextField(
                                                v_model=(f"params['{key}']",),
                                                density="compact",
                                                style="width: 70px",
                                                hide_details=True,
                                                single_line=True,
                                            )
            with v2.VRow():
                with v2.VCol():
                    with v2.VCard(style="width: 500px"):
                        with v2.VCardTitle("Output: a + b"):
                            with v2.VCardText():
                                v2.VTextField(
                                    v_model=("output",),
                                    density="compact",
                                    style="width: 70px",
                                    hide_details=True,
                                    #single_line=True,
                                    readonly=True,
                                )

if __name__ == "__main__":
    server.start()
```

Here's the output that I'm trying to get (with the output text field reactive to the inputs):
![Screenshot from 2024-09-26 17-52-07](https://github.com/user-attachments/assets/b6337cc2-18cd-4e58-8e9b-2d519aa77171)

In the code I'm simply trying to define two input parameters stored in a state dictionary, `state.params`, and compute their sum to be stored in a state output, `state.output`.

The reason for wanting to make this work with a dictionary is that I'm working on a more complex application where I need to read the number, name and initial value of such input parameters from an external file, without knowing those in advance.

#### Question 1
The first problem I encounter is that these lines
```py
# initialize output
params_array = np.array(list(state.params.values()))
state.output = model(params_array)
```
produce the following error:
```sh
Error: Skip state value for 'output' since its content is not serializable
```
Do you know what's wrong with defining `state.output` in this way, given that `model` simply returns a scalar?

#### Question 2
The second problem I encounter (when fixing manually the first one by setting, for example, `state.output = 3` so that the application does not crash) is that the function `params_react` does not get triggered by a change (through the slider or the text field box next to it) in the values of either of the two parameters stored in the dictionary `state.params`. When I define and initialize `state.params` I call `state.dirty("params")`. I think I might be missing a call to `state.flush()` somewhere, but I couldn't figure out where that call should go. Where should I `state.flush()` in order to detect changes coming from the slider or the text field box next to it?

Thanks a lot in advance!

**Comment by jourdain** [comment]
The serialization issue is caused by setting a numpy object onto the state. If you convert the result to a Python float it will be fine.

**Comment by jourdain** [comment]
The other issue is that you are editing a nested variable in the slider which means you are losing the reactivity client side unless you mark `params` [dirty](https://trame.readthedocs.io/en/latest/client.intro.html#flushState).

**Comment by jourdain** [comment]
```python
v2.VSlider(
      v_model=(f"params['{key}']",),
      change="flushState('params')",
)
```

**Comment by EZoni** [comment, marked as answer]
Thanks a lot! Converting the NumPy scalar to a Python float worked and adding `change="flushState('params')"` to the `VSlider` and `VTextField` elements fixed the reactivity issue.

**Comment by EZoni** [comment]
Just one more question on this, sorry.

Given that `state.output` is a read-only output variable, whose value changes due to the user interaction with other editable elements in the GUI (the input parameters), if `state.output` was a dictionary as well, say a dictionary with just one item `state.output["c"] = model(params_array)` in this case (as in, c=a+b), how would I tell Trame that the `VTextField` box showing `state.output["c"]` (via `v_model="output['c']"`) should react and display the updated value, whenever `state.output["c"]` changes due to the user interaction with other elements in the GUI (the input parameters)?

So basically are there commands equivalent to `state.dirty('params')` and `change=flushState('params')` when I want to obtain reactivity for a nested output variable whose items change because of editing other nested input variables? In other words, something to force push _from the server to the client_ when the server has updated one item of a state _nested_ object due to a change in other state nested variables, as oppose to force pushing from the client to the server (which I think is what `flushState` does).

I've tried `state.dirty('output')` and `change=flushState('output')` (in the output `VSlider`) but it doesn't seem to work, if I'm not mistaken. I also tried `state.flush()`, added manually after `state.output["c"]` is modified, but this also doesn't seem to work, meaning that the `VTextField` box showing `state.output["c"]` does not react. So maybe I'm missing something here.

**Reply by EZoni** [reply]
Some pseudo-code snippets to explain what I wrote:
```py
...
...
# initialize output dictionary (trivial example with only 1 item)
state.output = dict()
state.dirty("output")
state.output["c"] = model(params)
...
...
# react to parameters change (works)
@state.change("params")
def params_react(params, **kwargs):
    state.output["c"] = model(params)
    state.flush()
...
...
# display updated values in output dictionary (doesn't work)
v2.VTextField(
    v_model=("output['c']",),
    readonly=True,
    ...
...
...
```

**Comment by jourdain** [comment]
```python
@state.change("params")
def params_react(params, **kwargs):
    state.output["c"] = model(params)
    state.dirty("output")  # https://trame.readthedocs.io/en/latest/core.state.html#trame_server.state.State.dirty
...
...
# display updated values in output dictionary (doesn't work)
v2.VTextField(
    value=("output['c']",),
    readonly=True,
    ...
```

**Reply by EZoni** [reply]
Oh, I thought `state.dirty(...)` could be called only once and would then "mark" the variable through the rest of the code. Thanks!

**Comment by EZoni** [comment]
Hi @jourdain, one more quick follow-up question, about using `VTextField` to display numbers.

If I am dealing with numbers that are very large (as opposed to the `1`, `2` and `3` shown in the screenshots above), such as, for example, `123456789.123`, shall I use something other than `VTextField` to display them and be able to choose the format with which the numbers are displayed (e.g., scientific notation, exponential, etc.)?

For instance, I would like the number to be displayed as $1.23456789123e8$, or $1.23456789123 \text{x} 10^8$, or even better $1.23e8$, rather than $123456789.123$, which seems to be the only format I get with `VTextField`. Am I missing out how to customize the format with `VTextField` or should I use another widget?

Thanks in advance!

**Reply by jourdain** [reply]
If you set the `type="number"` and `v_model_number=...` it should automatically use those scientific notations. But otherwise something more close the JS may be needed to enable automatic formatting.

**Comment by EZoni** [comment]
Hi @jourdain, I am encountering another issue on an example similar to the one that we were discussing here.

If I migrate the code here from Vuetify 2 to Vuetify 3, the application, when launched, seems to stall indefinitely.

This is the new code using Vuetify 3:
```py
import numpy as np

from trame.app import get_server
from trame.ui.vuetify3 import SinglePageLayout
from trame.widgets import vuetify3 as vuetify

# toy model
def model(x_array):
    return float(np.sum(x_array))

server = get_server(client_type="vue3")
state, ctrl = server.state, server.controller

# initialize inputs
state.params = dict()
state.params["a"] = 2e5
state.params["b"] = 2e5
state.dirty("params")

# initialize output
params_array = np.array(list(state.params.values()))
state.output = dict()
state.output["c"] = model(params_array)
state.dirty("output")

# update output
@state.change("params")
def params_react(params, **kwargs):
    params_array = np.array(list(params.values()))
    print(f"\nparams_array = {params_array}")
    print(f"old output = {state.output}")
    state.output["c"] = model(params_array)
    print(f"new output = {state.output}")
    state.dirty("output")

@ctrl.add("center")
def center():
    state.params["a"] = 5e5
    state.params["b"] = 5e5
    state.dirty("params")

with SinglePageLayout(server) as layout:
    with layout.content:
        with vuetify.VContainer():
            with vuetify.VRow():
                with vuetify.VCol():
                    with vuetify.VCard(style="width: 500px"):
                        with vuetify.VCardTitle("Parameters"):
                            with vuetify.VCardText():
                                for key in state.params.keys():
                                    with vuetify.VSlider(
                                        v_model=(f"params['{key}']",),
                                        change="flushState('params')",
                                        label=f"{key}",
                                        min=0,
                                        max=10e5,
                                        step=(10e5-0)/100,
                                        hide_details=True,
                                        classes="align-center",
                                    ):
                                        with vuetify.Template(
                                            v_slot_append=True,
                                            __properties=[("v_slot_append", "v-slot:append")],
                                        ):
                                            vuetify.VTextField(
                                                v_model=(f"params['{key}']",),
                                                change="flushState('params')",
                                                density="compact",
                                                style="width: 70px",
                                                hide_details=True,
                                                single_line=True,
                                            )
                                vuetify.VSpacer(style="height: 20px")
                                vuetify.VBtn(
                                    "center",
                                    click=center,
                                )
            with vuetify.VRow():
                with vuetify.VCol():
                    with vuetify.VCard(style="width: 500px"):
                        with vuetify.VCardTitle("Output: a + b"):
                            with vuetify.VCardText():
                                vuetify.VTextField(
                                    v_model=("output['c']",),
                                    density="compact",
                                    style="width: 70px",
                                    hide_details=True,
                                    #single_line=True,
                                    readonly=True,
                                )

if __name__ == "__main__":
    server.start()
```

However, if you migrate back to Vuetify 2 by changing only the following lines
```py
from trame.ui.vuetify2 import SinglePageLayout
from trame.widgets import vuetify2 as vuetify
...
server = get_server(client_type="vue2")
```
the application, when launched, seems to run properly.

Do you have an idea of how to fix this particular issue?

FYI, I am encountering other issues with an application that builds on the building blocks described here, where the synchronization of parameters through flushing in sliders seems broken, and I wonder if I am missing something basic about migrating to Vuetify 3. I was hoping to be able to use the minimal working example in this discussion to debug my larger application, but then I realized that this example is broken too, as described above.

In case it helps, these are the trame packages in my current conda environment:
```console
$ conda list | grep trame
trame                     3.8.0              pyhd8ed1ab_0    conda-forge
trame-client              3.6.0              pyhd8ed1ab_0    conda-forge
trame-plotly              3.0.2              pyhd8ed1ab_1    conda-forge
trame-router              2.3.0              pyhd8ed1ab_1    conda-forge
trame-server              3.3.0              pyhd8ed1ab_0    conda-forge
trame-vuetify             2.8.1              pyhd8ed1ab_0    conda-forge
```

**Reply by jourdain** [reply]
You can not use v_model when using nested structure, you need to use `:valueModel` and `@update:valueModel`.

But since you are using vue3, you can use the new nested reactive infrastructure to simplify that usage. You can find an example [here](https://github.com/Kitware/trame-client/blob/master/examples/vue3/nested_reactive.py#L90).

**Reply by EZoni** [reply]
Just to clarify, does this mean that it should be enough to wrap things with a `with client.DeepReactive("my_structure") as dr:` or that it is necessary to rewrite everything using `html` instead of the `vuetify3` widgets?

**Reply by jourdain** [reply]
It should be enough to wrap things with DeepReactive and use v_model without any flushState trick.

**Reply by EZoni** [reply]
Thanks! For some weird reason I did not manage to make it work (i.e., avoid stalling when launching the application) for the minimal example posted above, but I did manage to make it work for the larger, more complex application that I cannot paste here. In that case wrapping things with `DeepReactive` was indeed enough to fix the reactivity with respect to parameters in a nested structure.
